llvm/lib/Target/AVR/MCTargetDesc/AVRAsmBackend.cpp

```diff
@@ -60,31 +70,37 @@ static void unsigned_width(unsigned Width, uint64_t Value,
 
     Ctx->reportError(Fixup.getLoc(), Diagnostic);
   }
+
+  return false;
 }
 
 /// Adjusts the value of a branch target before fixup application.
 static void adjustBranch(unsigned Size, const MCFixup &Fixup, uint64_t &Value,
                          MCContext *Ctx) {
   // We have one extra bit of precision because the value is rightshifted by
   // one.
-  unsigned_width(Size + 1, Value, std::string("branch target"), Fixup, Ctx);
+  checkUnsignedWidth(Size + 1, Value, std::string("branch target"), Fixup, Ctx);
 
   // Rightshifts the value by one.
   AVR::fixups::adjustBranchTarget(Value);
 }
 
 /// Adjusts the value of a relative branch target before fixup application.
-static void adjustRelativeBranch(unsigned Size, const MCFixup &Fixup,
-                                 uint64_t &Value, MCContext *Ctx) {
+static bool adjustRelativeBranch(unsigned Size, const MCFixup &Fixup,
+                                 uint64_t &Value, MCContext *Ctx,
+                                 const MCSubtargetInfo *STI = nullptr) {
   // Jumps are relative to the current instruction.
   Value -= 2;
 
   // We have one extra bit of precision because the value is rightshifted by
   // one.
   Size += 1;
 
-  if (!isIntN(Size, Value) &&
-      Ctx->getSubtargetInfo()->hasFeature(AVR::FeatureWrappingRjmp)) {
+  if (!STI) {
```

Comment:
In no context where there is a branch should the subtarget be uncertain?

Reply:
There's two contexts this function is called from:
- from within `adjustFixupValue()`, where we're gotten `MCContext *Ctx`,
- from within `shouldForceRelocation()`, where we're gotten only `const MCSubtargetInfo *STI`.

This `if (!STI)` part is invoked when we're being called from `adjustFixupValue()` and it just fetches `STI` from the `Ctx`, since it's available there. Alternatively we could adjust the call-sites of this function to be `(..., Ctx, Ctx->getSubtargetInfo())`, but I thought automatic substitution would be less error-prone.

Reply:
The subtarget is a function level concept, the global one shouldn't really exist. The subtarget should be passed in from the original callsite

Reply:
> The subtarget is a function level concept

I don't think so, you specify the subtarget when you run `llc`, it's also explicitly mentioned in `MCAsmBackendCtorTy` etc., no?

Reply:
This is a legacy issue, the subtarget should always come from the function

Reply:
I agree. It would be better that we only fix the relocation issue in this PR, and fix other  parts in other PRs.